infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/log/LogWatcher.java

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright (c) 2012-2018 Red Hat, Inc.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.workspace.infrastructure.kubernetes.namespace.log;
+
+import static org.eclipse.che.api.workspace.shared.Constants.DEBUG_WORKSPACE_START_LOG_LIMIT_BYTES;
+
+import io.fabric8.kubernetes.client.KubernetesClient;
+import java.io.Closeable;
+import java.util.Map;
+import java.util.Set;
+import java.util.StringJoiner;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.workspace.infrastructure.kubernetes.KubernetesClientFactory;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.event.PodEvent;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.event.PodEventHandler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class watches workspace's pod events and tries to read the logs of all its containers.
+ *
+ * <p>Current implementation uses provided thread-pool and each container log watch session runs in
+ * separate thread from this thread-pool.
+ *
+ * <p>Watching logs of individual containers is delegated to instances of {@link ContainerLogWatch}.
+ */
+public class LogWatcher implements PodEventHandler, Closeable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(LogWatcher.class);
+
+  // 10MB
+  public static final Long DEFAULT_LOG_LIMIT_BYTES = 10L * 1024 * 1024;
+
+  private static final String STARTED_EVENT_REASON = "Started";
+
+  private final KubernetesClient client;
+  private final Set<PodLogHandler> logHandlers = ConcurrentHashMap.newKeySet();
+  private final Executor containerWatchersThreadPool;
+  private final LogWatchTimeouts timeouts;
+  private final long inputStreamLimit;
+
+  private final String namespace;
+  private final String workspaceId;
+  private final Set<String> podsOfInterest;
+
+  private final AtomicBoolean closed = new AtomicBoolean(false);
```

Comment:
are you sure AtomicBoolean is needed here?

Reply:
I'm sure it's not. fixed